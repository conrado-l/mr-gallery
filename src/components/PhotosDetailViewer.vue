<template>
  <transition name="fade">
    <!-- Overlay -->
    <div v-if="visible" class="overlay">
      <!-- Outer container -->
      <div class="outer-container">
        <!-- Viewer container -->
        <div class="viewer-container">
          <div class="d-flex justify-content-center align-items-center w-100 h-100">
            <!-- Photo container -->
            <div class="photo-container"
                 :class="{zoom: isZooming, loading: isPhotoLoading }"
                 ref="photoContainer">
              <!-- Photo -->
              <img class="photo"
                   :src="getImageURLSource"
                   :key="getPhotoID"
                   draggable="false"
                   @load="onPhotoLoad()"
                   @click="zoomPhoto()"
                   @mousemove="handleMouseMove($event)"
                   @mousedown="handleMouseDown($event)"
                   @mouseup="handleMouseUp($event)">
            </div>
          </div>
          <!-- Footer -->
          <div class="footer">
            <!-- Title container -->
            <div class="title-container text-align-center">
              <span class="font-weight-600 font-color-white">{{getPhotoTitle}}</span>
            </div>

            <!-- Description container -->
            <div class="description-container text-align-center">
              <span class=font-color-white>{{getPhotoDescription}}</span>
            </div>
          </div>
        </div>
        <!-- Close button -->
        <div class="close-button-container cursor-pointer"
             title="Close viewer"
             @click="closeViewer()">
          <img class="overlay-action-button"
               src="../assets/images/icons/close.svg"
               >
        </div>
        <!-- Share button -->
        <div v-show="!isPhotoLoading"
             class="share-button-container cursor-pointer"
             title="Share photo URL"
             @click="onPhotoURLShare()">
          <img class="overlay-action-button"
               src="../assets/images/icons/share.svg">
        </div>
      </div>
    </div>
  </transition>
</template>

<script>
export default {
  name: 'PhotosDetailViewer',
  data: () => ({
    visible: false,
    isPhotoLoading: true,
    isZooming: false,
    isDragging: false,
    isSwipping: false,
    isDraggingSwipe: false,
    swipeType: null,
    scale: 1,
    top: 0,
    left: 0,
    lastX: 0,
    lastY: 0,
    canZoom: true,
    transition: 'all .3s ease',
    initialMouseX: 0,
    initialMouseY: 0,
    endMouseX: 0,
    endMouseY: 0
  }),
  props: {
    /** Contains the photos to be displayed **/
    photos: {
      type: Array,
      required: true
    },
    /** Photo's index for the current open photo **/
    currentPhotoIndex: {
      type: Number
    },
    /** Thumbnail object field name **/
    thumbnailField: {
      type: String,
      required: false,
      default: 'thumbnailPhoto'
    },
    /** Full photo object field name **/
    fullPhotoField: {
      type: String,
      required: false,
      default: 'fullPhoto'
    },
    /** Title photo object field name **/
    titleField: {
      type: String,
      required: false,
      default: 'title'
    },
    /** Description photo object field name **/
    descriptionField: {
      type: String,
      required: false,
      default: 'description'
    },
    /** Close the viewer if the backdrop is clicked **/
    backdropClose: {
      type: Boolean,
      required: false,
      default: true
    }
  },
  computed: {
    /**
     * Gets the thumbnail or full size image URL for the current open photo
     * @returns {string}
     */
    getImageURLSource () {
      const fullPhotoSrc = this.getPhotoField(this.fullPhotoField)

      // Full photo is available
      if (fullPhotoSrc) {
        return fullPhotoSrc
      } else {
        const thumbnailSrc = this.getPhotoField(this.thumbnailField)

        // Thumbnail fallback
        if (thumbnailSrc) {
          return thumbnailSrc
        } else {
          // Placeholder fallback if the full photo and thumbnail failed
          return 'https://demofree.sirv.com/nope-not-here.jpg'
        }
      }
    },
    /**
     * Gets the photo's title
     * @returns {string}
     */
    getPhotoTitle () {
      return this.getPhotoField(this.titleField) || ''
    },
    /**
     * Gets the photo's description
     * @returns {string}
     */
    getPhotoDescription () {
      return this.getPhotoField(this.descriptionField) || ''
    },
    /**
     * Gets the photo's id
     * @returns {string}
     */
    getPhotoID () {
      return this.getPhotoField('id') || ''
    },
    /**
     * Gets the photo's full size URL
     * @returns {string}
     */
    getPhotoFullSizeURL () {
      return this.getPhotoField(this.fullPhotoField) || ''
    },
    /**
     * Gets the photo's full size URL
     * @returns {string}
     */
    getPhotoThumbnailURL () {
      return this.getPhotoField(this.thumbnailField) || ''
    }
  },
  methods: {
    /**
     * Checks if there are photos available to visible
     * @returns {boolean}
     */
    arePhotosAvailable () {
      return (!!(this.photos && Array.isArray(this.photos) && this.photos.length))
    },
    /**
     * Notifies when the viewer is closed
     */
    closeViewer () {
      this.resetViewerSettings()
      this.$emit('close', this.currentPhotoIndex)
    },
    /**
     * Next photo
     */
    nextPhoto () {
      this.isPhotoLoading = true
      this.resetZoom()
      this.$emit('photo-changed', this.currentPhotoIndex + 1)
    },
    /**
     * Previous photo
     */
    previousPhoto () {
      const newCurrentPhotoIndex = this.currentPhotoIndex - 1

      // Check if the new photo is valid
      if (newCurrentPhotoIndex >= 0) {
        this.isPhotoLoading = true
        this.resetZoom()
        this.$emit('photo-changed', newCurrentPhotoIndex)
      }
    },
    /**
     * Enables/disables scrolling based on the viewer status (open/closed)
     */
    setScrolling () {
      if (this.visible) {
        this.disableScrolling()
      } else {
        this.enableScrolling()
      }
    },
    /**
     * Disables scrolling
     */
    disableScrolling () {
      document.body.setAttribute('style', 'overflow:hidden;')
    },
    /**
     * Enables scrolling
     */
    enableScrolling () {
      document.body.setAttribute('style', 'overflow:visible;')
    },
    /**
     * Register the key shortcuts
     */
    shortcutEventListener (event) {
      switch (event.keyCode) {
        // Right arrow
        case 39:
          return this.nextPhoto()
        // Left arrow
        case 37:
          return this.previousPhoto()
        // Prevent default down/up arrows
        case 38:
        case 40:
        case ' ':
          return event.preventDefault()
        // Escape key
        case 27:
          return this.closeViewer()
      }
    },
    /**
     * Gets a field from the current photo being displayed
     * @param {string} field
     * @return {string|number}
     */
    getPhotoField (field) {
      let photoField

      if (this.arePhotosAvailable()) {
        // Get the current open photo
        const currentOpenPhoto = this.photos[this.currentPhotoIndex]

        // Check if the current photo is valid
        if (currentOpenPhoto) {
          photoField = currentOpenPhoto[field]
        }
      }

      return photoField
    },
    /**
     * Handles the mouse move event
     */
    handleMouseMove (e) {
      if (!this.checkMouseEventPropButton(e.button)) {
        return
      }

      if (this.isDraging) {
        this.top = this.top - this.lastY + e.clientY
        this.left = this.left - this.lastX + e.clientX
        this.lastX = e.clientX
        this.lastY = e.clientY
        this.canZoom = false

        const item = e.target.parentNode
        const newZoom = 1.6
        item.style.transform = 'translate3d(calc(-50% + ' + this.left + 'px), calc(-50% + ' + this.top + 'px), 0px) scale3d(' + newZoom + ', ' + newZoom + ', ' + newZoom + ')'
      }
      e.stopPropagation()
    },
    /**
     * Handles the mouse down event
     */
    handleMouseDown (e) {
      if (!this.checkMouseEventPropButton(e.button)) {
        return
      }

      this.lastX = e.clientX
      this.lastY = e.clientY
      this.isDraging = true
      e.stopPropagation()
    },
    /**
     * Handles the mouse up event
     */
    handleMouseUp (e) {
      if (!this.checkMouseEventPropButton(e.button)) {
        return
      }

      this.isDraging = false

      this.lastX = 0
      this.lastY = 0

      const thisContext = this
      setTimeout(function () {
        thisContext.canZoom = true
      }, 100)
    },
    /**
     * Checks if it's allowed to drag
     */
    checkMouseEventPropButton (button) {
      if (!this.isZooming) {
        return false
      }

      return button === 0
    },

    /**
     * Zooms on the photo
     */
    zoomPhoto () {
      if (!this.canZoom) {
        return false
      }

      if (this.isSwipping) {
        return false
      }

      const item = this.$refs.photoContainer

      const isZooming = this.isZooming
      const thisContext = this

      if (isZooming) {
        if (!this.isDraging) {
          this.isZooming = false
        }
      } else {
        this.isZooming = true
      }
      if (this.isZooming) {
        item.style.transform = 'translate3d(calc(-50%), calc(-50%), 0px) scale3d(1.6, 1.6, 1.6)'
        setTimeout(function () {
          thisContext.transition = 'all .0s ease'
        }, 100)
      } else {
        this.resetZoom()
      }
    },
    /**
     * Resets the zoom
     */
    resetZoom () {
      this.scale = 1
      this.left = 0
      this.top = 0
      this.isZooming = false
      this.swipeType = null
      this.transition = 'all .3s ease'

      if (this.currentPhotoIndex != null) {
        const item = this.$refs.photoContainer

        item.style.transform = 'translate3d(calc(-50% + ' + this.left + 'px), calc(-50% + ' + this.top + 'px), 0px) scale3d(1, 1, 1)'
        this.initialMouseX = 0
      }
    },
    /**
     * Resets the viewer settings to default
     */
    resetViewerSettings () {
      this.visible = false
      this.resetZoom()
      this.isDragging = false
      this.isPhotoLoading = true
    },
    /**
     * Sets the loading indicator when the photo is loaded
     */
    onPhotoLoad () {
      if (this.isFullPhotoURLSet()) {
        this.isPhotoLoading = false
      }
    },
    /**
     * Checks if the full size photo is currently set as source
    */
    isFullPhotoURLSet () {
      return !!this.getPhotoFullSizeURL
    },
    /**
     * Closes the viewer when the backdrop is clicked
     */
    onBackdropClick () {
      if (this.backdropClose) {
        this.closeViewer()
      }
    },
    /** Shares the full photo URL
     * Closes the viewer when the backdrop is clicked
     */
    onPhotoURLShare () {
      if (!this.isPhotoLoading) {
        this.copyTextToClipboard(this.getImageURLSource)
      }
    },
    /**
     * Copies a text to the clipboard
     */
    copyTextToClipboard (text) {
      const el = document.createElement('textarea')
      el.value = text
      document.body.appendChild(el)
      el.select()
      document.execCommand('copy')
      document.body.removeChild(el)
    }
  },
  watch: {
    /**
     * Checks if the viewer needs to be opened, closed or change the photo
     * @param {number} newValue
     * @param {number} oldValue
     */
    currentPhotoIndex (newValue, oldValue) {
      // TODO: find way to avoid retriggering in a loop
      if (newValue === oldValue) {
        return
      }

      // Show/hide the viewer
      this.visible = newValue != null

      // Enable/disable scrolling
      this.setScrolling()
    }
  },
  mounted () {
    // Register the keydown event listener for key shortcuts
    window.addEventListener('keydown', this.shortcutEventListener)
  },
  beforeDestroy () {
    // Remove the keydown event listener for key shortcuts
    window.removeEventListener('keydown', this.shortcutEventListener)

    // Enable scrolling
    this.enableScrolling()
  }
}
</script>

<style lang="scss" scoped>

@import "../assets/styles/helpers";

.overlay {
  position: fixed;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 2;
}

.outer-container {
  position: relative;
}

.viewer-container {
  position: relative;
  width: 100vw;
  height: 100vh;
  z-index: 3;
}

.photo-container {
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate3d(calc(-50% + 0px), calc(-50% + 0px), 0px) scale3d(1, 1, 1);
  cursor: zoom-in;
  box-shadow: 0 0 1.5rem rgba(0, 0, 0, .45);
  z-index: 5;

  &.zoomed {
    cursor: zoom-out;
  }

  &.loading {
    cursor: wait;
  }
}

.photo {
  margin: auto;
}

.overlay-action-button {
  width: 25px;
  height: 25px;
}

.close-button-container {
  position: absolute;
  right: 15px;
  top: 15px;
  z-index: 4;
}

.share-button-container {
  position: absolute;
  right: 65px;
  top: 15px;
  z-index: 4;
}

.footer {
  position: absolute;
  bottom: 0;
  width: 100%;
  padding: 20px;
}

.title-container {
  bottom: 5vh;
}

.description-container {
  bottom: 3vh;
}

// Animations
.fade-enter-active, .fade-leave-active {
  transition: opacity .3s;
}
.fade-enter, .fade-leave-to {
  opacity: 0;
}

</style>
