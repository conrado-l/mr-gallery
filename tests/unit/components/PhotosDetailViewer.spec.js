// Vue Test Utils
import { createLocalVue, mount } from '@vue/test-utils'

// Components
import PhotosDetailViewer from '@/components/PhotosDetailViewer'

// Mocks
import photosMock from './mocks/photos'

describe('PhotosDetailViewer.vue', () => {
  const localVue = createLocalVue()

  // Default props
  const defaultProps = {
    photos: []
  }

  // Used selectors
  const selectors = {
    viewer: '[data-test="viewer-container"]',
    mainPhoto: '[data-test="main-photo"]',
    photoContainer: '[data-test="photo-container"]',
    nextPhotoButton: '[data-test="next-photo-button"]',
    previousPhotoButton: '[data-test="previous-photo-button"]',
    sharePhotoButton: '[data-test="share-url-button"]'
  }

  // Reusable factory mount
  const factoryMount = (props, attachToDocument = false) => {
    return mount(PhotosDetailViewer, {
      localVue,
      propsData: {
        ...props
      },
      attachToDocument
    })
  }

  it.skip('should render the component correctly', () => {
    const wrapper = factoryMount(defaultProps)

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render the thumbnail photo since there is no full size one', async () => {
    const photos = [
      {
        id: 'd0d3fb4d11aaee9accce',
        thumbnailPhoto: 'http://interview.agileengine.com/pictures/cropped/0019.jpg'
      }
    ]
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos,
      currentPhotoIndex: 0,
      useThumbnail: true
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Check if the thumbnail photo was rendered
    expect(mainPhotoWrapper.element.src).toBe(photos[0].thumbnailPhoto)
  })

  it('should render the full size photo', async () => {
    const photos = [
      {
        id: 'd0d3fb4d11aaee9accce',
        thumbnailPhoto: 'http://interview.agileengine.com/pictures/cropped/0019.jpg',
        fullPhoto: 'http://interview.agileengine.com/pictures/full_size/00219.jpg'
      }
    ]
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos,
      currentPhotoIndex: 0
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Check if the full size photo was rendered
    expect(mainPhotoWrapper.element.src).toBe(photos[0].fullPhoto)
  })

  // TODO: research why "return require('@/assets/images/image-not-found.webp')" is breaking the test
  it.skip('should render the fallback/placeholder error photo', async () => {
    const photos = [
      {
        id: 'd0d3fb4d11aaee9accce'
      }
    ]
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos,
      currentPhotoIndex: 0
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Check if the fallback error photo was rendered
    expect(mainPhotoWrapper.element.src).toBe(require('../../../src/assets/images/image-not-found.webp'))
  })

  it('should render the previous photo after clicking the left navigation arrow', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock,
      currentPhotoIndex: 1
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the previous photo button
    const previousPhotoButtonWrapper = wrapper.find(selectors.previousPhotoButton)

    // Click on the button
    await previousPhotoButtonWrapper.trigger('click')

    // Check that the event was emitted for the new photo index, being 0
    expect(wrapper.emitted('photo-changed')[0]).toEqual([0])

    // Set the new index photo
    wrapper.setProps({
      currentPhotoIndex: 0
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Check if the full size photo was rendered
    expect(mainPhotoWrapper.element.src).toBe(photosMock[0].fullPhoto)
  })

  it('should render the next photo after pressing the left arrow key', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock,
      currentPhotoIndex: 1
    }, true)

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Press the right arrow key
    await wrapper.trigger('keydown.left')

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Check that the event was emitted for the new photo index, being 0
    expect(wrapper.emitted('photo-changed')[0]).toEqual([0])

    // Set the new index photo
    wrapper.setProps({
      currentPhotoIndex: 0
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Check if the full size photo was rendered
    expect(mainPhotoWrapper.element.src).toBe(photosMock[0].fullPhoto)

    // Destroy the wrapper since we called attachToDocument
    wrapper.destroy()
  })

  it('should render the next photo after clicking the right navigation arrow', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock,
      currentPhotoIndex: 0
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the next photo button
    const nextPhotoButtonWrapper = wrapper.find(selectors.nextPhotoButton)

    // Click on the button
    await nextPhotoButtonWrapper.trigger('click')

    // Check that the event was emitted for the new photo index, being 1
    expect(wrapper.emitted('photo-changed')[0]).toEqual([1])

    // Set the new index photo
    wrapper.setProps({
      currentPhotoIndex: 1
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Check if the full size photo was rendered
    expect(mainPhotoWrapper.element.src).toBe(photosMock[1].fullPhoto)
  })

  it('should render the next photo after pressing the right arrow key', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock,
      currentPhotoIndex: 0
    }, true)

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Press the right arrow key
    await wrapper.trigger('keydown.right')

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Check that the event was emitted for the new photo index, being 1
    expect(wrapper.emitted('photo-changed')[0]).toEqual([1])

    // Set the new index photo
    wrapper.setProps({
      currentPhotoIndex: 1
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Check if the full size photo was rendered
    expect(mainPhotoWrapper.element.src).toBe(photosMock[1].fullPhoto)

    // Destroy the wrapper since we called attachToDocument
    wrapper.destroy()
  })

  it('should enable the zoom on the photo when clicking on it', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock,
      currentPhotoIndex: 0
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Click on it
    await mainPhotoWrapper.trigger('click')

    // Find the photo container
    const photoContainerWrapper = wrapper.find(selectors.photoContainer)

    // Check if the photo is zoomed
    expect(photoContainerWrapper.element.style.transform).toBe('translate3d(calc(-50%), calc(-50%), 0px) scale3d(2, 2, 2)')
  })

  it('should disable the zoom on the photo when clicking on it twice', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock,
      currentPhotoIndex: 0
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the main photo
    const mainPhotoWrapper = wrapper.find(selectors.mainPhoto)

    // Click on it to enable the zoom
    await mainPhotoWrapper.trigger('click')

    // Click on it to disable the zoom
    await mainPhotoWrapper.trigger('click')

    // Find the photo container
    const photoContainerWrapper = wrapper.find(selectors.photoContainer)

    // Check if the photo is zoomed
    expect(photoContainerWrapper.element.style.transform).toBe('translate3d(calc(-50% + 0px), calc(-50% + 0px), 0px) scale3d(1, 1, 1)')
  })

  // TODO: mock document.execCommand / copyTextToClipboard
  it.skip('should show a toast notification after clicking the share button', async () => {
    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos: photosMock,
      currentPhotoIndex: 0
    })

    wrapper.vm.copyTextToClipboard = jest.fn()

    // Set the loading state to false
    wrapper.setData({ isPhotoLoading: false })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the share photo button
    const sharePhotoButtonWrapper = wrapper.find(selectors.sharePhotoButton)

    // Click on the button
    await sharePhotoButtonWrapper.trigger('click')

    // Find the toast notification
    const toastNotificationWrapper = wrapper.find('.v-toast-text')

    // Check if the toast was shown with the correct message
    expect(toastNotificationWrapper.text()).toBe('The photo URL was copied to the clipboard')
  })
})
