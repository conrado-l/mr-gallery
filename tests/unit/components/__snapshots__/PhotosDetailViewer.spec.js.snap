// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotosDetailViewer.vue should render the component correctly 1`] = `
<transition-stub
  name="fade"
>
  <!---->
</transition-stub>
`;
